k8s-pb-codegen/src/main.rs

```diff
@@ -95,5 +107,172 @@ fn main() -> Result<()> {
+        .open(target_dir.join("lib.rs"))
+        .unwrap();
+    append_trait_def(&mut lib_rs);
+
     Ok(())
 }
+
+fn append_trait_def(lib_rs: &mut File) {
+    let tokens = quote! {
+        pub trait Resource {
+            const API_VERSION: &'static str;
+            const GROUP: &'static str;
+            const KIND: &'static str;
+            const VERSION: &'static str;
+            const NAME: &'static str;
+        }
+
+        pub trait HasMetadata {
+            type Metadata;
+            fn metadata(&self) -> Option<&Self::Metadata>;
+            fn metadata_mut(&mut self) -> Option<&mut Self::Metadata>;
+        }
+        pub trait HasSpec {
+            type Spec;
+            fn spec(&self) -> Option<&Self::Spec>;
+            fn spec_mut(&mut self) -> Option<&mut Self::Spec>;
+        }
+        pub trait HasStatus {
+            type Status;
+            fn status(&self) -> Option<&Self::Status>;
+            fn status_mut(&mut self) -> Option<&mut Self::Status>;
+        }
+        pub trait HasConditions {
+            type Condition;
+            fn conditions(&self) -> Option<&[Self::Condition]>;
+            fn conditions_mut(&mut self) -> Option<&mut Vec<Self::Condition>>;
+        }
+    };
+    let tokens = rustfmt(tokens).unwrap();
+    writeln!(lib_rs).unwrap();
+    writeln!(lib_rs, "{}", &tokens).unwrap();
+}
+
+fn append_trait_impl(pkg_rs: &mut File, message_name: &str, resource: &Resource) {
+    use heck::CamelCase;
+    // Convert to match prost
+    let type_name = message_name.to_camel_case();
+    let type_name = format_ident!("{}", type_name);
+
+    let api_version = &resource.api_version;
+    let group = &resource.group;
+    let kind = &resource.kind;
+    let version = &resource.version;
+    let name = &resource.name;
+    let tokens = quote! {
+        impl crate::Resource for #type_name {
+            const API_VERSION: &'static str = #api_version;
+            const GROUP: &'static str = #group;
+            const KIND: &'static str = #kind;
+            const VERSION: &'static str = #version;
+            const NAME: &'static str = #name;
+        }
+    };
+    let tokens = if let Some(metadata) = &resource.metadata {
+        let metadata = metadata.split("::").map(|m| format_ident!("{}", m));
+        quote! {
+            #tokens
+
+            impl crate::HasMetadata for #type_name {
+                type Metadata = crate::#(#metadata)::*;
+
+                fn metadata(&self) -> Option<&<Self as crate::HasMetadata>::Metadata> {
+                    self.metadata.as_ref()
+                }
+                fn metadata_mut(&mut self) -> Option<&mut<Self as crate::HasMetadata>::Metadata> {
+                    self.metadata.as_mut()
+                }
+            }
+        }
+    } else {
+        tokens
+    };
+    let tokens = if let Some(spec) = &resource.spec {
+        // HACK Prost changes message name `APIService` to `ApiService`, and `transformed.json` uses the original name.
```

Comment:
Prost uses `heck` to transform message name, so some of them doesn't match the name extracted from `swagger.json`.

Reply:
That is nicer, I think. Although it creates the problem of us having to disambiguate `APIResource` with `ApiResource`, but that might be something we ought to resolve in another way.

Reply:
Yeah, makes them fit Rust's naming convention better. But I'm not sure if we should start using different casing from the official ones. We can worry about it later.

Reply:
yeah, i'll open an issue to track it